Compilation error in AudioService.java (Flutter plugin audio_service 0.18.15) when running ./gradlew build --refresh-dependencies

### Minimal reproduction project
Official example: main.dart
### Build command
```
./gradlew build --refresh-dependencies
```
### Build error
```
F:\flutter\.pub-cache\hosted\pub.dev\audio_service-0.18.15\android\src\main\java\com\ryanheise\audioservice\AudioService.java:389:
Error: Must be one of: Service.STOP_FOREGROUND_LEGACY, Service.STOP_FOREGROUND_REMOVE, Service.STOP_FOREGROUND_DETACH [WrongConstant]
              stopForeground(removeNotification ? STOP_FOREGROUND_REMOVE : 0);
```
### Output of flutter doctor
```
[√] Flutter (Channel stable, 3.24.3, on Microsoft Windows [Version 10.0.19045.4894], locale pl-PL)
    • Flutter version 3.24.3 on channel stable at C:\src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (3 weeks ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at C:\Android\android-sdk
    • Platform android-35, build-tools 34.0.0
    • ANDROID_HOME = C:\Android\android-sdk
    • Java binary at: C:\Users\user\AppData\Local\Programs\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 21.0.3+-12282718-b509.11)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Build Tools 2019 16.11.27)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools
    • Visual Studio Build Tools 2019 version 16.11.33801.447
    • Windows 10 SDK version 10.0.22621.0

[!] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    X Unable to determine bundled Java version.
    • Try updating or re-installing Android Studio.

[√] Android Studio (version 2024.1)
    • Android Studio at C:\Users\user\AppData\Local\Programs\Android Studio 2
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.11+0--11852314)

[√] Android Studio (version 2024.2)
    • Android Studio at C:\Users\user\AppData\Local\Programs\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.3+-12282718-b509.11)

[√] VS Code (version 1.93.1)
    • VS Code at C:\Users\user\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.98.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.4894]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 129.0.6668.89
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 128.0.2739.42

[√] Network resources
    • All expected network resources are available.
```

same here version 0.18.17 looks like related to **lint** file

> Execution failed for task ':audio_service:lintDebug'.
> > Lint found errors in the project; aborting build.
>   
>   Fix the issues identified by lint, or create a baseline to see only new errors.
>   To create a baseline, run gradlew updateLintBaseline after adding the following to the module's build.gradle file:
>   
> android {
>       lint {
>           baseline = file("lint-baseline.xml")
>       }
>   }
>   
>   /Users//.pub-cache/hosted/pub.dev/audio_service-0.18.17/android/src/main/java/com/ryanheise/audioservice/AudioService.java:389: Error: Must be one of: Service.STOP_FOREGROUND_LEGACY, Service.STOP_FOREGROUND_REMOVE, Service.STOP_FOREGROUND_DETACH [WrongConstant]
>               stopForeground(removeNotification ? STOP_FOREGROUND_REMOVE : 0);

please update this:

> stopForeground(removeNotification ? STOP_FOREGROUND_REMOVE : 0);

to this or similar:

> stopForeground(removeNotification ? STOP_FOREGROUND_REMOVE : STOP_FOREGROUND_DETACH);

--- a/.gitignore
+++ b/.gitignore
@@ -47,6 +47,7 @@ gradlew
 gradlew.bat
 local.properties
 **/GeneratedPluginRegistrant.java
+**/.cxx/
 
 # Android Studio will place build artifacts here
 **/android/app/debug

--- a/audio_service/CHANGELOG.md
+++ b/audio_service/CHANGELOG.md
@@ -1,6 +1,7 @@
 ## 0.18.18
 
 * Fix setPlaybackState entitlement issue on iOS.
+* Fix Android stopForeground deprecation warning.
 * Bump compile/target SDK to 35 on Android.
 * Bump minSdk to 19 on Android.
 * bump AGP to 8.5.2.

--- a/audio_service/CHANGELOG.md
+++ b/audio_service/CHANGELOG.md
@@ -1,6 +1,9 @@
 ## 0.18.18
 
 * Fix setPlaybackState entitlement issue on iOS.
+* Bump compile/target SDK to 35 on Android.
+* Bump minSdk to 19 on Android.
+* bump AGP to 8.5.2.
 
 ## 0.18.17
 

--- a/audio_service/android/build.gradle
+++ b/audio_service/android/build.gradle
@@ -3,13 +3,16 @@ version "1.0-SNAPSHOT"
 def args = ["-Xlint:deprecation","-Xlint:unchecked"]
 
 buildscript {
+    ext {
+        agp_version = '8.5.2'
+    }
     repositories {
         google()
         mavenCentral()
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:8.5.2'
+        classpath "com.android.tools.build:gradle:$agp_version"
     }
 }
 
@@ -27,15 +30,12 @@ project.getTasks().withType(JavaCompile) {
 apply plugin: "com.android.library"
 
 android {
-    // Conditional for compatibility with AGP <4.2.
-    if (project.android.hasProperty("namespace")) {
-        namespace 'com.ryanheise.audioservice'
-    }
-    compileSdkVersion 35
+    namespace "com.ryanheise.audioservice"
+    compileSdk = 35
 
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_11
-        targetCompatibility JavaVersion.VERSION_11
+        sourceCompatibility = JavaVersion.VERSION_11
+        targetCompatibility = JavaVersion.VERSION_11
     }
     defaultConfig {
         minSdk = 19

--- a/audio_service/android/build.gradle
+++ b/audio_service/android/build.gradle
@@ -1,5 +1,5 @@
-group 'com.ryanheise.audioservice'
-version '1.0-SNAPSHOT'
+group = "com.ryanheise.audioservice"
+version "1.0-SNAPSHOT"
 def args = ["-Xlint:deprecation","-Xlint:unchecked"]
 
 buildscript {
@@ -9,7 +9,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.3.0'
+        classpath 'com.android.tools.build:gradle:8.5.2'
     }
 }
 
@@ -24,23 +24,21 @@ project.getTasks().withType(JavaCompile) {
     options.compilerArgs.addAll(args)
 }
 
-apply plugin: 'com.android.library'
+apply plugin: "com.android.library"
 
 android {
     // Conditional for compatibility with AGP <4.2.
     if (project.android.hasProperty("namespace")) {
         namespace 'com.ryanheise.audioservice'
     }
-    compileSdkVersion 33
+    compileSdkVersion 35
 
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_1_8
-        targetCompatibility JavaVersion.VERSION_1_8
+        sourceCompatibility JavaVersion.VERSION_11
+        targetCompatibility JavaVersion.VERSION_11
     }
     defaultConfig {
-        minSdkVersion 16
-        targetSdkVersion 33
-        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+        minSdk = 19
     }
     lintOptions {
         disable 'InvalidPackage'
@@ -48,5 +46,6 @@ android {
 }
 
 dependencies {
-    implementation 'androidx.media:media:1.6.0'
+    implementation 'androidx.media:media:1.7.0'
+    implementation 'androidx.core:core:1.13.1'
 }

--- a/audio_service/android/gradle/wrapper/gradle-wrapper.properties
+++ b/audio_service/android/gradle/wrapper/gradle-wrapper.properties
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.7-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-all.zip

--- a/audio_service/android/src/main/AndroidManifest.xml
+++ b/audio_service/android/src/main/AndroidManifest.xml
@@ -1,3 +1,2 @@
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-  package="com.ryanheise.audioservice">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
 </manifest>

--- a/audio_service/android/src/main/java/com/ryanheise/audioservice/AudioService.java
+++ b/audio_service/android/src/main/java/com/ryanheise/audioservice/AudioService.java
@@ -29,6 +29,7 @@ import android.util.Size;
 import android.view.KeyEvent;
 
 import androidx.annotation.RequiresApi;
+import androidx.core.app.ServiceCompat;
 import androidx.core.content.ContextCompat;
 import androidx.core.app.NotificationCompat;
 import androidx.media.MediaBrowserServiceCompat;
@@ -371,7 +372,7 @@ public class AudioService extends MediaBrowserServiceCompat {
         artBitmapCache.evictAll();
         compactActionIndices = null;
         releaseMediaSession();
-        legacyStopForeground(!config.androidResumeOnClick);
+        ServiceCompat.stopForeground(this, config.androidResumeOnClick ? STOP_FOREGROUND_DETACH : STOP_FOREGROUND_REMOVE);
         // This still does not solve the Android 11 problem.
         // if (notificationCreated) {
         //     NotificationManager notificationManager = getNotificationManager();
@@ -380,18 +381,6 @@ public class AudioService extends MediaBrowserServiceCompat {
         releaseWakeLock();
         instance = null;
         notificationCreated = false;
-    }
-
-    @SuppressWarnings("deprecation")
-    private void legacyStopForeground(boolean removeNotification) {
-        if (Build.VERSION.SDK_INT >= 24) {
-            // TODO: Consider application of STOP_FOREGROUND_DETACH
-            stopForeground(removeNotification ? STOP_FOREGROUND_REMOVE : 0);
-        } else {
-            // TODO: This API is deprecated and we'll need to eventually
-            // delete this line.
-            stopForeground(removeNotification);
-        }
     }
 
     public AudioServiceConfig getConfig() {
@@ -730,7 +719,7 @@ public class AudioService extends MediaBrowserServiceCompat {
     }
 
     private void exitForegroundState() {
-        legacyStopForeground(false);
+        ServiceCompat.stopForeground(this, STOP_FOREGROUND_DETACH);
         releaseWakeLock();
     }
 

--- a/audio_service/example/android/app/build.gradle
+++ b/audio_service/example/android/app/build.gradle
@@ -1,44 +1,33 @@
-def localProperties = new Properties()
-def localPropertiesFile = rootProject.file('local.properties')
-if (localPropertiesFile.exists()) {
-    localPropertiesFile.withReader('UTF-8') { reader ->
-        localProperties.load(reader)
-    }
+plugins {
+    id "com.android.application"
+    id "kotlin-android"
+    // The Flutter Gradle Plugin must be applied after the Android and Kotlin Gradle plugins.
+    id "dev.flutter.flutter-gradle-plugin"
 }
-
-def flutterRoot = localProperties.getProperty('flutter.sdk')
-if (flutterRoot == null) {
-    throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
-}
-
-def flutterVersionCode = localProperties.getProperty('flutter.versionCode')
-if (flutterVersionCode == null) {
-    flutterVersionCode = '1'
-}
-
-def flutterVersionName = localProperties.getProperty('flutter.versionName')
-if (flutterVersionName == null) {
-    flutterVersionName = '1.0'
-}
-
-apply plugin: 'com.android.application'
-apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
 
 android {
-    namespace 'com.ryanheise.audioserviceexample'
-    compileSdkVersion 34
+    namespace = "com.ryanheise.audioserviceexample"
+    compileSdk = 35
 
-    lintOptions {
-        disable 'InvalidPackage'
+    compileOptions {
+        sourceCompatibility = JavaVersion.VERSION_11
+        targetCompatibility = JavaVersion.VERSION_11
+    }
+
+    kotlinOptions {
+        jvmTarget = JavaVersion.VERSION_11
+    }
+
+    lint {
+        disable = ["InvalidPackage"]
     }
 
     defaultConfig {
         applicationId "com.ryanheise.audioserviceexample"
-        minSdkVersion 21
-        targetSdkVersion 34
-        versionCode flutterVersionCode.toInteger()
-        versionName flutterVersionName
-        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+        minSdk = flutter.minSdkVersion
+        targetSdk = 35
+        versionCode = flutter.versionCode
+        versionName = flutter.versionName
     }
 
     buildTypes {

--- a/audio_service/example/android/app/build.gradle
+++ b/audio_service/example/android/app/build.gradle
@@ -7,7 +7,7 @@ plugins {
 
 android {
     namespace = "com.ryanheise.audioserviceexample"
-    compileSdk = 35
+    compileSdk = flutter.compileSdkVersion
 
     compileOptions {
         sourceCompatibility = JavaVersion.VERSION_11
@@ -18,14 +18,10 @@ android {
         jvmTarget = JavaVersion.VERSION_11
     }
 
-    lint {
-        disable = ["InvalidPackage"]
-    }
-
     defaultConfig {
         applicationId "com.ryanheise.audioserviceexample"
         minSdk = flutter.minSdkVersion
-        targetSdk = 35
+        targetSdk = flutter.targetSdkVersion
         versionCode = flutter.versionCode
         versionName = flutter.versionName
     }
@@ -37,6 +33,10 @@ android {
             signingConfig signingConfigs.debug
             shrinkResources true
         }
+    }
+
+    lint {
+        disable = ["InvalidPackage"]
     }
 }
 

--- a/audio_service/example/android/app/src/debug/AndroidManifest.xml
+++ b/audio_service/example/android/app/src/debug/AndroidManifest.xml
@@ -1,5 +1,4 @@
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.ryanheise.audioserviceexample">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
     <!-- Flutter needs it to communicate with the running application
          to allow setting breakpoints, to provide hot reload, etc.
     -->

--- a/audio_service/example/android/app/src/main/AndroidManifest.xml
+++ b/audio_service/example/android/app/src/main/AndroidManifest.xml
@@ -1,6 +1,5 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:tools="http://schemas.android.com/tools"
-    package="com.ryanheise.audioserviceexample">
+    xmlns:tools="http://schemas.android.com/tools">
 
     <uses-permission android:name="android.permission.INTERNET"/>
     <uses-permission android:name="android.permission.WAKE_LOCK"/>

--- a/audio_service/example/android/build.gradle
+++ b/audio_service/example/android/build.gradle
@@ -1,14 +1,3 @@
-buildscript {
-    repositories {
-        google()
-        mavenCentral()
-    }
-
-    dependencies {
-        classpath 'com.android.tools.build:gradle:7.3.0'
-    }
-}
-
 allprojects {
     repositories {
         google()
@@ -16,12 +5,12 @@ allprojects {
     }
 }
 
-rootProject.buildDir = '../build'
+rootProject.buildDir = "../build"
 subprojects {
     project.buildDir = "${rootProject.buildDir}/${project.name}"
 }
 subprojects {
-    project.evaluationDependsOn(':app')
+    project.evaluationDependsOn(":app")
 }
 
 tasks.register("clean", Delete) {

--- a/audio_service/example/android/gradle.properties
+++ b/audio_service/example/android/gradle.properties
@@ -2,3 +2,6 @@ org.gradle.jvmargs=-Xmx1536M
 android.enableJetifier=true
 android.useAndroidX=true
 android.enableR8=true
+android.defaults.buildfeatures.buildconfig=true
+android.nonTransitiveRClass=false
+android.nonFinalResIds=false

--- a/audio_service/example/android/gradle/wrapper/gradle-wrapper.properties
+++ b/audio_service/example/android/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,6 @@
-#Fri Jun 23 08:50:38 CEST 2017
+#Sat Mar 22 15:01:44 AEDT 2025
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.7-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-all.zip

--- a/audio_service/example/android/settings.gradle
+++ b/audio_service/example/android/settings.gradle
@@ -1,15 +1,25 @@
-include ':app'
+pluginManagement {
+    def flutterSdkPath = {
+        def properties = new Properties()
+        file("local.properties").withInputStream { properties.load(it) }
+        def flutterSdkPath = properties.getProperty("flutter.sdk")
+        assert flutterSdkPath != null, "flutter.sdk not set in local.properties"
+        return flutterSdkPath
+    }()
 
-def flutterProjectRoot = rootProject.projectDir.parentFile.toPath()
+    includeBuild("$flutterSdkPath/packages/flutter_tools/gradle")
 
-def plugins = new Properties()
-def pluginsFile = new File(flutterProjectRoot.toFile(), '.flutter-plugins')
-if (pluginsFile.exists()) {
-    pluginsFile.withReader('UTF-8') { reader -> plugins.load(reader) }
+    repositories {
+        google()
+        mavenCentral()
+        gradlePluginPortal()
+    }
 }
 
-plugins.each { name, path ->
-    def pluginDirectory = flutterProjectRoot.resolve(path).resolve('android').toFile()
-    include ":$name"
-    project(":$name").projectDir = pluginDirectory
+plugins {
+    id "dev.flutter.flutter-plugin-loader" version "1.0.0"
+    id "com.android.application" version "8.5.2" apply false
+    id "org.jetbrains.kotlin.android" version "1.8.22" apply false
 }
+
+include ":app"

--- a/audio_service/example/pubspec.yaml
+++ b/audio_service/example/pubspec.yaml
@@ -4,7 +4,6 @@ publish_to: 'none'
 
 environment:
   sdk: ">=3.3.0 <4.0.0"
-  flutter: ">=3.0.0"
 
 dependencies:
   flutter:

--- a/audio_service/lib/audio_service.dart
+++ b/audio_service/lib/audio_service.dart
@@ -3318,9 +3318,10 @@ mixin QueueHandler on BaseAudioHandler {
   }
 
   @override
-  Future<void> updateQueue(List<MediaItem> newQueue) async {
-    queue.add(queue.nvalue!..replaceRange(0, queue.nvalue!.length, newQueue));
-    await super.updateQueue(newQueue);
+  Future<void> updateQueue(List<MediaItem> queue) async {
+    this.queue.add(
+        this.queue.nvalue!..replaceRange(0, this.queue.nvalue!.length, queue));
+    await super.updateQueue(queue);
   }
 
   @override

--- a/audio_service/pubspec.yaml
+++ b/audio_service/pubspec.yaml
@@ -8,8 +8,8 @@ topics:
   - background
 
 environment:
-  sdk: ">=3.3.0 <4.0.0"
-  flutter: ">=3.0.0"
+  sdk: ^3.6.0
+  flutter: ">=3.27.0"
 
 dependencies:
   # Use these deps for local development

--- a/audio_service/pubspec.yaml
+++ b/audio_service/pubspec.yaml
@@ -34,7 +34,7 @@ dependencies:
   audio_service_platform_interface: ^0.1.3
   audio_service_web: ^0.1.4
 
-  audio_session: ^0.1.25
+  audio_session: '>=0.1.25 <0.3.0'
   rxdart: '>=0.26.0 <0.29.0'
   flutter_cache_manager: ^3.3.1
   clock: ^1.1.0

